File: scripts/rust/src/review_struct_file_naming.py
#!/usr/bin/env python3
"""
Review: Struct naming must match file name.

Finds cases where a struct name doesn't match the file name pattern.

Example violation:
  File: src/Chap05/RelationStEph.rs
  Contains: pub struct Relation<A, B>  // Should be RelationStEph

Rule: The primary struct in a file should have the same base name as the file.
For files like FooStEph.rs, the struct should be FooStEph, not just Foo.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "lib"))
from review_utils import ReviewContext, create_review_parser


def check_file(file_path: Path, context: ReviewContext) -> list[str]:
    """
    Check if struct names match the file name pattern.
    Returns list of violation messages.
    """
    violations = []
    
    try:
        content = file_path.read_text()
        lines = content.split('\n')
        
        # Get the expected struct name from file name
        # e.g., RelationStEph.rs -> RelationStEph
        file_stem = file_path.stem  # Remove .rs extension
        
        # Pattern to match struct declarations
        # Matches: pub struct Foo, pub struct Foo<T>, pub(crate) struct Foo, etc.
        struct_pattern = re.compile(r'^\s*pub(?:\([^)]*\))?\s+struct\s+(\w+)')
        
        for line_num, line in enumerate(lines, 1):
            match = struct_pattern.match(line)
            if match:
                struct_name = match.group(1)
                
                # Check if struct name matches file name
                if struct_name != file_stem:
                    # Also check if it's the "S" suffix variant (FooS vs Foo)
                    if struct_name + 'S' == file_stem or struct_name == file_stem + 'S':
                        continue  # This is acceptable (FooS struct in Foo.rs or vice versa)
                    
                    rel_path = context.relative_path(file_path)
                    violations.append(
                        f"  {rel_path}:{line_num} - struct '{struct_name}' doesn't match file name '{file_stem}'\n"
                        f"    {line.strip()}"
                    )
    
    except Exception as e:
        print(f"Error reading {file_path}: {e}", file=sys.stderr)
    
    return violations


def main():
    parser = create_review_parser(
        description="Check that struct names match file names (RustRules.md - Naming conventions)"
    )
    args = parser.parse_args()
    
    context = ReviewContext(args)
    all_violations = []
    
    # Search in src/, tests/, and benches/
    for directory in ['src', 'tests', 'benches']:
        dir_path = context.repo_root / directory
        if not dir_path.exists():
            continue
        
        rust_files = list(dir_path.rglob('*.rs'))
        
        for file_path in rust_files:
            if args.file and context.relative_path(file_path) != args.file:
                continue
            
            violations = check_file(file_path, context)
            all_violations.extend(violations)
    
    if all_violations:
        print("✗ Struct/File Naming violations found:\n", file=sys.stderr)
        for violation in all_violations:
            print(violation, file=sys.stderr)
        print(f"\nTotal violations: {len(all_violations)}\n", file=sys.stderr)
        print("Struct names should match their file names (excluding .rs extension).", file=sys.stderr)
        return 1
    else:
        if not args.file:
            print("✓ Struct/File Naming: No violations found (RustRules.md)")
        else:
            print(f"✓ {args.file}: No violations found")
        return 0


if __name__ == '__main__':
    sys.exit(main())

File: src/bin/review_struct_file_naming.rs
// Copyright (C) Brian G. Milnes 2025

//! Review: Struct naming must match file name
//! 
//! Replaces: scripts/rust/src/review_struct_file_naming.py
//! Rule: The primary struct in a file should have the same base name as the file
//! Example: FooStEph.rs should contain `pub struct FooStEph`, not just `pub struct Foo`
//! Git commit: 584a672b6a34782766863c5f76a461d3297a741a
//! Binary: rusticate-review-struct-file-naming

use anyhow::Result;
use rusticate::{StandardArgs, format_number, parse_source, find_nodes, find_rust_files};
use ra_ap_syntax::{SyntaxKind, ast::{self, AstNode}};
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_struct_file_naming.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
#[derive(Debug)]
struct Violation {
    file: PathBuf,
    line_num: usize,
    struct_name: String,
    file_stem: String,
}

fn check_file(file_path: &Path, source: &str) -> Result<Vec<Violation>> {
    let source_file = parse_source(source)?;
    let root = source_file.syntax();
    
    // Get expected struct name from file name
    let file_stem = file_path
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("")
        .to_string();
    
    // Find all public struct definitions
    let struct_nodes = find_nodes(root, SyntaxKind::STRUCT);
    let mut violations = Vec::new();
    
    for struct_node in struct_nodes {
        if let Some(struct_ast) = ast::Struct::cast(struct_node.clone()) {
            // Check if struct has pub visibility
            let has_pub = struct_node.siblings_with_tokens(rowan::Direction::Prev)
                .any(|s| s.to_string().contains("pub"));
            
            if !has_pub {
                continue; // Skip private structs
            }
            
            // Get struct name from children
            let struct_name = struct_node.children_with_tokens()
                .filter_map(|child| child.into_token())
                .find(|token| token.kind() == SyntaxKind::IDENT)
                .map(|token| token.text().to_string());
            
            if let Some(struct_name) = struct_name {
                
                // Check if name matches file stem
                if struct_name != file_stem {
                    // Allow "S" suffix variant (FooS vs Foo)
                    let name_with_s = format!("{struct_name}S");
                    let stem_with_s = format!("{file_stem}S");
                    
                    if name_with_s != file_stem && struct_name != stem_with_s {
                        let line_num = rusticate::line_number(struct_ast.syntax(), source);
                        violations.push(Violation {
                            file: file_path.to_path_buf(),
                            line_num,
                            struct_name,
                            file_stem: file_stem.clone(),
                        });
                    }
                }
            }
        }
    }
    
    Ok(violations)
}

fn main() -> Result<()> {
    let start = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();
    
    // Print compilation directory for Emacs compile-mode
    log!("Entering directory '{}'", base_dir.display());
    log!("");
    
    let search_dirs = args.get_search_dirs();
    let files = find_rust_files(&search_dirs);
    let mut all_violations = Vec::new();
    
    for file in &files {
        match fs::read_to_string(file) {
            Ok(source) => {
                match check_file(file, &source) {
                    Ok(violations) => all_violations.extend(violations),
                    Err(e) => {
                        eprintln!("Warning: Failed to parse {}: {}", file.display(), e);
                    }
                }
            }
            Err(e) => {
                eprintln!("Warning: Failed to read {}: {}", file.display(), e);
            }
        }
    }
    
    // Report violations
    if all_violations.is_empty() {
        log!("✓ Struct/File Naming: No violations found");
    } else {
        log!("✗ Struct/File Naming violations found:");
        log!("");
        
        for v in &all_violations {
            if let Ok(rel_path) = v.file.strip_prefix(&base_dir) {
                log!("{}:{}: struct '{}' doesn't match file name '{}'", 
                         rel_path.display(), v.line_num, v.struct_name, v.file_stem);
            }
        }
        
        log!("");
        log!("Struct names should match their file names (excluding .rs extension).");
    }
    
    // Summary
    log!("");
    log!("Summary: {} files checked, {} violations", 
             format_number(files.len()), format_number(all_violations.len()));
    
    let elapsed = start.elapsed().as_millis();
    log!("Completed in {}ms", elapsed);
    
    if all_violations.is_empty() {
        Ok(())
    } else {
        std::process::exit(1);
    }
}

